Drivers/Src/CANbus.c

```diff
@@ -3,206 +3,241 @@
 }
 
+//wrapper functions for the interrupt customized for each bus
+void CANbus_TxHandler_1(){
+    CANbus_TxHandler(CAN_1);
+}
+
+void CANbus_RxHandler_1(){
+    CANbus_RxHandler(CAN_1);
+}
+void CANbus_TxHandler_3(){
+    CANbus_TxHandler(CAN_3);
+}
+void CANbus_RxHandler_3(){
+    CANbus_RxHandler(CAN_3);
+}
+
 /**
- * @brief   Initializes the CAN system
- * @param   None
- * @return  None
+ * @brief   Initializes the CAN system for a given bus. Must be called independently for each bus.
+ * @param   bus
  */
-void CANbus_Init(void)
+void CANbus_Init(CAN_t bus)
 {
     // initialize CAN mailbox semaphore to 3 for the 3 CAN mailboxes that we have
     // initialize tx
     OS_ERR err;
-
-    OSMutexCreate(&CANbus_TxMutex, "CAN TX Lock", &err);
+    
+    OSMutexCreate(&(CANbus_TxMutex[bus]), (bus == CAN_1 ? "CAN TX Lock 1":"CAN TX Lock 3"), &err);
     assertOSError(OS_CANDRIVER_LOC,err);
 
-    OSMutexCreate(&CANbus_RxMutex, "CAN RX Lock", &err);
+    OSMutexCreate(&(CANbus_RxMutex[bus]), (bus == CAN_1 ? "CAN RX Lock 1":"CAN RX Lock 3"), &err);
     assertOSError(OS_CANDRIVER_LOC,err);
 
-    OSSemCreate(&CANMail_Sem4, "CAN Mailbox Semaphore", 3, &err); // there's 3 hardware mailboxes on the board, so 3 software mailboxes
+    OSSemCreate(&(CANMail_Sem4[bus]), (bus == CAN_1 ? "CAN Mailbox Semaphore 1":"CAN Mailbox Semaphore 3"), 3, &err); // there's 3 hardware mailboxes on the board, so 3 software mailboxes
     assertOSError(OS_CANDRIVER_LOC,err);
 
-    OSSemCreate(&CANBus_RecieveSem4, "CAN Recieved Msg queue", 0, &err); // create a mailbox counter to hold the messages in as they come in
+    OSSemCreate(&(CANBus_RecieveSem4[bus]), (bus == CAN_1 ? "CAN Recieved Msg Queue Ctr 1":"CAN Recieved Msg Queue Ctr 3"), 0, &err); // create a mailbox counter to hold the messages in as they come in
     assertOSError(OS_CANDRIVER_LOC,err);
 
-    BSP_CAN_Init(CAN_1, &CANbus_RxHandler, &CANbus_TxHandler);
+    if(bus==CAN_1){
+        BSP_CAN_Init(bus,&CANbus_RxHandler_1,&CANbus_TxHandler_1);
+    } else if (bus==CAN_3){
+        BSP_CAN_Init(bus,&CANbus_RxHandler_3,&CANbus_TxHandler_3);
+    }
+
 }
 
 /**
- * @brief   Transmits data onto the CANbus
+ * @brief   Transmits data onto the specified CANbus. Transmits up to 8 bytes at a time. If more is necessary, please use an IDX message.
  * @param   id : CAN id of the message
  * @param 	payload : the data that will be sent.
- * @param blocking: Whether or not this should be a blocking call
- * @return  0 if data wasn't sent, otherwise it was sent.
+ * @param   blocking: Whether or not this should be a blocking call
+ * @return  ERROR if data wasn't sent, SUCCESS if it was sent.
  */
-ErrorStatus CANbus_Send(CANId_t id, CANPayload_t payload, CAN_blocking_t blocking)
+ErrorStatus CANbus_Send(CANDATA_t CanData,CAN_blocking_t blocking, CAN_t bus)
 {
     CPU_TS timestamp;
     OS_ERR err;
+
+    CANLUT_T msginfo = CANLUT[CanData.ID]; //lookup msg information in table
```

Comment:
could crash for invalid IDs
need to validate ID < num IDs before accessing

also, there needs to be some way to see if IDs are valid (check for can len zero value?)

Reply:
👍